### ISSUE TITLE
Card is not charged. 

### ISSUE THREAD

**Issue by chuksdsilent** [Flutterwave/rave-android#226]
I am trying to implement ravenonuimanager but the card is not charged. How do I go about it. thanks

Please help my I am finding it very difficult implementing this to my app thanks.

my code.

public class StripePayment extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener, CardPaymentCallback   {
       CardPayManager cardPayManager;
      Card card

        pay.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                transactionID = "tranxav_" +
                        generateRandomValues(30);

                Preferences.setDefaults("transactionID", transactionID, getApplicationContext());
                makePayment(100);
  
            }
        });

         public void makePayment(int amount){

            RaveNonUIManager raveNonUIManager =
                    new RaveNonUIManager()
                            .setAmount(100)
                            .setTxRef("ksldjfi")
                            .setEmail("my@email.com")
                            .setPublicKey("FLWPUBK_TEST-2f6f9eca9903a888c5fdc8982514cd20-X")
                            .setEncryptionKey("FLWSECK_TEST97c64bd835da")
                            .setTxRef("skdfwf")
                            .setPhoneNumber("08032547454")
                    .initialize();

             cardPayManager = new CardPaymentManager(raveNonUIManager, this );
             card = new Card("5531886652142950", "12", "22", "123"); // Test MasterCard PIN

            AlertDialog.Builder builder = new AlertDialog.Builder(StripePayment.this);
            builder.setTitle("Enter Pin");

            // Set up the input
            final EditText input = new EditText(StripePayment.this);

            // Specify the type of input expected; this, for example, sets the input as a password, and will mask the text
            input.setInputType(InputType.TYPE_CLASS_TEXT);
            builder.setView(input);

            // Set up the buttons
            builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    pin = input.getText().toString();

                    AlertDialog.Builder builders = new AlertDialog.Builder(StripePayment.this);
                    builders.setTitle("Enter Otp");

                    // Set up the input
                    final EditText otpInput = new EditText(StripePayment.this);

                    // Specify the type of input expected; this, for example, sets the input as a password, and will mask the text
                    otpInput.setInputType(InputType.TYPE_CLASS_TEXT);
                    builders.setView(otpInput);

                    // Set up the buttons
                    builders.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            otp = input.getText().toString();
                            cardPayManager.chargeCard(card);

                        }
                    });
                    builders.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.cancel();
                        }
                    });
                    builders.show();

                }
            });
            builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.cancel();
                }
            });

            builder.show();
        }

    @Override
    public void collectCardPin() {

        //collect PIN from User and pass to the CardPayManager object.
        cardPayManager.submitPin(pin);
    }

    @Override
    public void collectOtp(String message) {

        //collect OTP from User and pass to the CardPayManager object.
        cardPayManager.submitOtp(otp);
    }

    @Override
    public void onSuccessful(String flwRef) {
        Log.d(null, "onSuccessful: " + "The card was charged");
    }
}

**Comment by jeremiahAI** [comment]
Hi @chuksdsilent , You seem to be collecting the pin and OTP from the user after calling the charge endpoint, and not as a response to the `collectPin()` and `collectOtp()` callbacks. This will not yield a proper flow, as it is currently a race between the user's typing speed and the network speed (if the charge call is completed before the user enters the pin, an empty or null pin value will be sent back).

You need to collect the pin when the `collectPin` callback is called and call `cardPayManager.submitPin()` only after the PIN has been entered successfully. Same with the OTP. 

Please see the documentation on how to use our [charge utils](https://github.com/Flutterwave/rave-android/blob/master/ChargeVerificationUtils.md) to handle this. You can also clone the repo and check out the sample app's implementation.
